src/tools/rbd_mirror/Replayer.cc

```diff
@@ -770,36 +810,82 @@ bool Replayer::stop_image_replayer(unique_ptr<ImageReplayer<> > &image_replayer)
   return false;
 }
 
-void Replayer::handle_post_acquire_leader(Context *on_finish) {
+void Replayer::stop_image_replayers() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+  for (auto image_it = m_image_replayers.begin();
+       image_it != m_image_replayers.end();) {
+    if (stop_image_replayer(image_it->second)) {
+      image_it = m_image_replayers.erase(image_it);
+      continue;
+    }
+    ++image_it;
+  }
+}
+
+void Replayer::stop_image_replayers(Context *on_finish) {
   dout(20) << dendl;
 
   {
     Mutex::Locker locker(m_lock);
-    m_cond.Signal();
+    stop_image_replayers();
+
+    if (!m_image_replayers.empty()) {
+      Context *ctx = new FunctionContext([this, on_finish](int r) {
+          assert(r == 0);
+          stop_image_replayers(on_finish);
+        });
+      ctx = create_async_context_callback(m_threads->work_queue, ctx);
+      m_threads->timer->add_event_after(1, ctx);
```

Comment:
@dillaman Observing teuthology failures [1]

It looks like all cases are due to the timer lock is not held here:
```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.0-1436-gb9f5564/rpm/el7/BUILD/ceph-12.0.0-1436-gb9f5564/src/common/Timer.cc: 127: FAILED assert(lock.is_locked())
 ceph version 12.0.0-1436-gb9f5564 (b9f556438d3c7612e9c0a042c8d8f0959cabd3a0)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x110) [0x7f351bd9c1a0]
 2: (()+0x2b840d) [0x7f351bd9440d]
 3: (rbd::mirror::Replayer::stop_image_replayers(Context*)+0x179) [0x7f3524d19119]
 4: (rbd::mirror::Replayer::handle_shut_down_pool_watcher(int, Context*)+0xb6) [0x7f3524d194a6]
 5: (FunctionContext::finish(int)+0x2a) [0x7f3524d1dc0a]
 6: (Context::complete(int)+0x9) [0x7f3524d1cc29]
 7: (ThreadPool::worker(ThreadPool::WorkThread*)+0xb59) [0x7f351bda4649]
 8: (ThreadPool::WorkThread::entry()+0x10) [0x7f351bda5660]
 9: (()+0x7dc5) [0x7f351a686dc5]
 10: (clone()+0x6d) [0x7f351914573d]
```
[1] http://pulpito.ceph.com/trociny-2017-03-16_07:53:00-rbd-wip-mgolub-testing---basic-smithi/